{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bbce7a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Author: Janssen Kotah\n",
    "# snRNAseq analysis for WT/SHIP1 KO mice as part of Matera et al. project\n",
    "\n",
    "# Scrublet cleanup of SoupX processed sample, adjusting lab pipeline based on the following code from:\n",
    "# https://cellgeni.github.io/notebooks/html/new-doublets-scrublet.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3ecc92dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.io\n",
    "import os\n",
    "import scrublet as scr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0f1fd61e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.9.1 anndata==0.8.0 umap==0.5.3 numpy==1.23.5 scipy==1.9.3 pandas==1.5.3 scikit-learn==1.2.0 statsmodels==0.13.5 python-igraph==0.10.2 pynndescent==0.5.8\n"
     ]
    }
   ],
   "source": [
    "sc.settings.verbosity = 3\n",
    "sc.logging.print_header()\n",
    "sc.settings.set_figure_params(dpi=800, facecolor=\"white\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e0f90b20",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running scrublet for JK7\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.50\n",
      "Detected doublet rate = 0.6%\n",
      "Estimated detectable doublet fraction = 13.0%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 4.3%\n",
      "Elapsed time: 0.8 seconds\n",
      "Saving results for JK7\n",
      "Running scrublet for JK4\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.43\n",
      "Detected doublet rate = 2.2%\n",
      "Estimated detectable doublet fraction = 44.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 4.9%\n",
      "Elapsed time: 6.1 seconds\n",
      "Saving results for JK4\n",
      "Running scrublet for JK5\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.38\n",
      "Detected doublet rate = 2.8%\n",
      "Estimated detectable doublet fraction = 40.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 7.0%\n",
      "Elapsed time: 8.4 seconds\n",
      "Saving results for JK5\n",
      "Running scrublet for JK8\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.41\n",
      "Detected doublet rate = 2.2%\n",
      "Estimated detectable doublet fraction = 38.9%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 5.7%\n",
      "Elapsed time: 5.0 seconds\n",
      "Saving results for JK8\n",
      "Running scrublet for JK6\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.41\n",
      "Detected doublet rate = 2.5%\n",
      "Estimated detectable doublet fraction = 34.2%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 7.2%\n",
      "Elapsed time: 6.2 seconds\n",
      "Saving results for JK6\n",
      "Running scrublet for JK1\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.46\n",
      "Detected doublet rate = 2.5%\n",
      "Estimated detectable doublet fraction = 30.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 8.4%\n",
      "Elapsed time: 9.7 seconds\n",
      "Saving results for JK1\n",
      "Running scrublet for JK2\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.75\n",
      "Detected doublet rate = 0.1%\n",
      "Estimated detectable doublet fraction = 6.1%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 2.2%\n",
      "Elapsed time: 9.6 seconds\n",
      "Saving results for JK2\n",
      "Running scrublet for JK3\n",
      "--> This might be very slow. Consider passing `cache=True`, which enables much faster reading from a cache file.\n",
      "Preprocessing...\n",
      "Simulating doublets...\n",
      "Embedding transcriptomes using PCA...\n",
      "Calculating doublet scores...\n",
      "Automatically set threshold at doublet score = 0.35\n",
      "Detected doublet rate = 3.3%\n",
      "Estimated detectable doublet fraction = 46.5%\n",
      "Overall doublet rate:\n",
      "\tExpected   = 10.0%\n",
      "\tEstimated  = 7.0%\n",
      "Elapsed time: 12.5 seconds\n",
      "Saving results for JK3\n"
     ]
    }
   ],
   "source": [
    "samples_list = os.listdir(\"/data/bcn/p310674/Documents/009_Ship1_Paolicelli/04. Analysis_resequencing_2/001_SoupX_outputs_for_scrublet\")\n",
    "\n",
    "for sample_name in samples_list:\n",
    "    current_sample = sample_name[:3]\n",
    "    print(\"Running scrublet for \" + current_sample)\n",
    "    \n",
    "    input_dir = \"/data/bcn/p310674/Documents/009_Ship1_Paolicelli/04. Analysis_resequencing_2/001_SoupX_outputs_for_scrublet/\" + sample_name + \"/\"\n",
    "\n",
    "    adata = sc.read_10x_mtx(input_dir, cache = False)\n",
    "    adata.var_names_make_unique()\n",
    "    \n",
    "    scrub = scr.Scrublet(adata.X, expected_doublet_rate= 0.1)\n",
    "    adata.obs['doublet_scores'], adata.obs['predicted_doublets'] = scrub.scrub_doublets(min_counts = 2, \n",
    "                                                                                        min_cells = 3, \n",
    "                                                                                        min_gene_variability_pctl=85,\n",
    "                                                                                        n_prin_comps=30)\n",
    "    \n",
    "    print(\"Saving results for \" + current_sample)\n",
    "    path_save = current_sample + \"_scrublet_results.tsv\"\n",
    "    pd.DataFrame(adata.obs).to_csv(path_save, sep = '\\t', header = True)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
